providers/implementations/kdfs/hmacdrbg_kdf.c

```diff
@@ -0,0 +1,221 @@
+    ossl_unused void *vctx, ossl_unused void *p_ctx)
+{
+    static const OSSL_PARAM known_gettable_ctx_params[] = {
+        OSSL_PARAM_utf8_string(OSSL_KDF_PARAM_MAC, NULL, 0),
+        OSSL_PARAM_utf8_string(OSSL_KDF_PARAM_DIGEST, NULL, 0),
+        OSSL_PARAM_END
+    };
+    return known_gettable_ctx_params;
+}
+
+static int hmac_drbg_kdf_set_ctx_params(void *vctx,
+                                        const OSSL_PARAM params[])
+{
+    KDF_HMAC_DRBG *hmac = (KDF_HMAC_DRBG *)vctx;
+    PROV_DRBG_HMAC *drbg = &hmac->base;
+    OSSL_LIB_CTX *libctx = PROV_LIBCTX_OF(hmac->provctx);
+    const EVP_MD *md;
+    const OSSL_PARAM *p;
+    void *ptr = NULL;
+    size_t size = 0;
+
+    if (params == NULL)
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_HMACDRBG_ENTROPY);
+    if (p != NULL) {
+        if (!OSSL_PARAM_get_octet_string(p, &ptr, 0, &size))
+            return 0;
+        OPENSSL_free(hmac->entropy);
+        hmac->entropy = ptr;
+        hmac->entropylen = size;
+        hmac->init = 0;
+        ptr = NULL;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_HMACDRBG_NONCE);
+    if (p != NULL) {
+        if (!OSSL_PARAM_get_octet_string(p, &ptr, 0, &size))
+            return 0;
+        OPENSSL_free(hmac->nonce);
+        hmac->nonce = ptr;
+        hmac->noncelen = size;
+        hmac->init = 0;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_ALG_PARAM_DIGEST);
+    if (p != NULL) {
+        if (!ossl_prov_digest_load_from_params(&drbg->digest, params, libctx))
+            return 0;
+
+        /* Confirm digest is allowed. Allow all digests that are not XOF */
+        md = ossl_prov_digest_md(&drbg->digest);
+        if (md != NULL) {
+            if ((EVP_MD_get_flags(md) & EVP_MD_FLAG_XOF) != 0) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_XOF_DIGESTS_NOT_ALLOWED);
+                return 0;
+            }
+            drbg->blocklen = EVP_MD_get_size(md);
+        }
+        return ossl_prov_macctx_load_from_params(&drbg->ctx, params,
+                                                 "HMAC", NULL, NULL, libctx);
+    }
+    return 1;
+}
+
+static const OSSL_PARAM *hmac_drbg_kdf_settable_ctx_params(
+    ossl_unused void *vctx, ossl_unused void *p_ctx)
+{
+    static const OSSL_PARAM known_settable_ctx_params[] = {
+        OSSL_PARAM_octet_string(OSSL_KDF_PARAM_HMACDRBG_ENTROPY, NULL, 0),
+        OSSL_PARAM_octet_string(OSSL_KDF_PARAM_HMACDRBG_NONCE, NULL, 0),
+        OSSL_PARAM_utf8_string(OSSL_KDF_PARAM_DIGEST, NULL, 0),
+        OSSL_PARAM_utf8_string(OSSL_KDF_PARAM_PROPERTIES, NULL, 0),
+        OSSL_PARAM_END
+    };
+    return known_settable_ctx_params;
+}
+
+const OSSL_DISPATCH ossl_kdf_hmac_drbg_functions[] = {
+    { OSSL_FUNC_KDF_NEWCTX, (void(*)(void))hmac_drbg_kdf_new },
+    { OSSL_FUNC_KDF_FREECTX, (void(*)(void))hmac_drbg_kdf_free },
+    { OSSL_FUNC_KDF_RESET, (void(*)(void))hmac_drbg_kdf_reset },
```

Comment:
OSSL_FUNC_KDF_DUPCTX function is missing. We have it for all other KDFs.

Reply:
Not sure this is really necessary - but added anyway so it doesnt stall.. 
Added extra tests because dup is really easy to get wrong.

Reply:
I don't think it's strictly necessary but nice to have.